metric_map.go

```diff
@@ -0,0 +1,229 @@
+package gostatsd
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// MetricMap is used for storing aggregated or consolidated Metric values.
+// The keys of each map are metric names.
+type MetricMap struct {
+	Counters Counters
+	Timers   Timers
+	Gauges   Gauges
+	Sets     Sets
+}
+
+func NewMetricMap() *MetricMap {
+	return &MetricMap{
+		Counters: Counters{},
+		Timers:   Timers{},
+		Gauges:   Gauges{},
+		Sets:     Sets{},
+	}
+}
+
+// Receive adds a single Metric to the MetricMap, and releases the Metric.
+func (mm *MetricMap) Receive(m *Metric, now time.Time) {
+	tagsKey := m.TagsKey
+	nowNano := Nanotime(now.UnixNano())
+
+	switch m.Type {
+	case COUNTER:
+		mm.receiveCounter(m, tagsKey, nowNano)
+	case GAUGE:
+		mm.receiveGauge(m, tagsKey, nowNano)
+	case TIMER:
+		mm.receiveTimer(m, tagsKey, nowNano)
+	case SET:
+		mm.receiveSet(m, tagsKey, nowNano)
+	default:
+		logrus.StandardLogger().Errorf("Unknown metric type %s for %s", m.Type, m.Name)
+	}
+	m.Done()
+}
+
+func (mm *MetricMap) Merge(mmFrom *MetricMap) {
+	mmFrom.Counters.Each(func(metricName string, tagsKey string, counterFrom Counter) {
+		v, ok := mm.Counters[metricName]
+		if ok {
```

Comment:
consider handling the !ok case first and returning early to avoid all these nested ifs

Reply:
and for the functions following this one also